app/models/wpcc/your_contribution.rb

```diff
@@ -4,5 +4,20 @@ class YourContribution
     attr_accessor :eligible_salary,
                   :employee_percent,
                   :employer_percent
+
+    CONTRIBUTION_PREFERENCES = {
+      'minimum' => MinimumContributionCalculator,
+      'full' => FullContributionCalculator
+    }.freeze
+
+    def self.generate(salary, contribution_preference)
+      contribution_klass = CONTRIBUTION_PREFERENCES[contribution_preference]
+      calculator = contribution_klass.new(salary, contribution_preference)
+      new(
+        eligible_salary:  calculator.eligible_salary,
+        employee_percent: calculator.employee_percent,
+        employer_percent: calculator.employer_percent
+      )
+    end
```

Comment:
I understand the point, but the second one is instantiating itself. And it is 5 lines long due to line length limitation, otherwise will be only 3 lines.

@willhall88 & @margOnline  Is this solution sounds more agreeable?:

```ruby
  def self.generate(salary, contribution_preference)
      calculator = calculator_for(contribution_preference: contribution_preference)
      new(
        eligible_salary:  calculator.eligible_salary,
        employee_percent: calculator.employee_percent,
        employer_percent: calculator.employer_percent
      )
end

     def calculator_for(contribution_preference: contribution_preference)
       contribution_klass = CONTRIBUTION_PREFERENCES[contribution_preference]
       contribution_klass.new(salary, contribution_preference)
     end
# ...
```